tests/func/parsing/test_matrix.py

```diff
@@ -50,3 +50,47 @@ def test_matrix_interpolated(tmp_dir, dvc, matrix):
         "build@linux-3.8-dict1-list0": {"cmd": "echo linux 3.8 --arg2 2 -- out1 out11"},
         "build@linux-3.8-dict1-list1": {"cmd": "echo linux 3.8 --arg2 2 -- out2 out22"},
     }
+
+
+@pytest.mark.parametrize(
+    "matrix",
+    [
+        MATRIX_DATA,
+        {
+            "os": "${os}",
+            "pyv": "${pyv}",
+            "dict": "${dict}",
+            "list": "${list}",
+        },
+    ],
+)
+def test_matrix_key_present(tmp_dir, dvc, matrix):
+    (tmp_dir / "params.yaml").dump(MATRIX_DATA)
+    resolver = DataResolver(dvc, tmp_dir.fs_path, {})
+    data = {
+        "matrix": matrix,
+        "cmd": "echo ${key}",
+    }
+    definition = MatrixDefinition(resolver, resolver.context, "build", data)
+
+    assert definition.resolve_all() == {
+        f"build@{key}": {"cmd": f"echo {key}"}
+        for key in (
+            "win-3.7-dict0-list0",
+            "win-3.7-dict0-list1",
+            "win-3.7-dict1-list0",
+            "win-3.7-dict1-list1",
+            "win-3.8-dict0-list0",
+            "win-3.8-dict0-list1",
+            "win-3.8-dict1-list0",
+            "win-3.8-dict1-list1",
+            "linux-3.7-dict0-list0",
+            "linux-3.7-dict0-list1",
+            "linux-3.7-dict1-list0",
+            "linux-3.7-dict1-list1",
+            "linux-3.8-dict0-list0",
+            "linux-3.8-dict0-list1",
+            "linux-3.8-dict1-list0",
+            "linux-3.8-dict1-list1",
```

Comment:
Do you mind unrolling this loop, please? That might make it easier to see when things fail.

Reply:
Done